server/src/main/java/io/druid/segment/realtime/firehose/ReplayableFirehoseFactory.java

```diff
@@ -0,0 +1,313 @@
+{
+  private static final EmittingLogger log = new EmittingLogger(ReplayableFirehoseFactory.class);
+  private static final boolean DELETE_TEMP_FILES = true;
+  private static final boolean DEFAULT_REPORT_PARSE_EXCEPTIONS = false;
+  private static final int DEFAULT_MAX_TEMP_FILE_SIZE = 250000000;
+  private static final int DEFAULT_READ_FIREHOSE_RETRIES = 3;
+
+  private final FirehoseFactory delegateFactory;
+  private final boolean reportParseExceptions;
+  private final int maxTempFileSize;
+  private final int readFirehoseRetries;
+
+  private ReplayableFirehose firehose;
+
+  @JsonCreator
+  public ReplayableFirehoseFactory(
+      @JsonProperty("delegate") FirehoseFactory delegateFactory,
+      @JsonProperty("reportParseExceptions") Boolean reportParseExceptions,
+      @JsonProperty("maxTempFileSize") Integer maxTempFileSize,
+      @JsonProperty("readFirehoseRetries") Integer readFirehoseRetries
+  )
+  {
+    Preconditions.checkNotNull(delegateFactory, "delegate cannot be null");
+    Preconditions.checkArgument(
+        !(delegateFactory instanceof ReplayableFirehoseFactory),
+        "Refusing to wrap another ReplayableFirehoseFactory"
+    );
+
+    this.delegateFactory = delegateFactory;
+    this.reportParseExceptions = reportParseExceptions == null
+                                 ? DEFAULT_REPORT_PARSE_EXCEPTIONS
+                                 : reportParseExceptions;
+    this.maxTempFileSize = maxTempFileSize == null ? DEFAULT_MAX_TEMP_FILE_SIZE : maxTempFileSize;
+    this.readFirehoseRetries = readFirehoseRetries == null ? DEFAULT_READ_FIREHOSE_RETRIES : readFirehoseRetries;
+
+    log.info(this.toString());
+  }
+
+  @Override
+  public Firehose connect(InputRowParser parser) throws IOException
+  {
+    if (firehose == null) {
+      firehose = new ReplayableFirehose(parser);
+    } else {
+      log.info("Rewinding and returning existing firehose");
+      firehose.rewind();
+    }
+
+    return firehose;
+  }
+
+  @Override
+  public void close() throws IOException
+  {
+    if (firehose != null) {
+      firehose.possiblyDeleteTempFiles();
+      firehose.close();
+    }
+
+    firehose = null;
+  }
+
+  public class ReplayableFirehose implements Firehose
+  {
+    private final JsonFactory factory = new DefaultObjectMapper(
+        new SmileFactory().enable(SmileGenerator.Feature.CHECK_SHARED_STRING_VALUES)
+    ).getFactory();
+
+    private final List<File> files = Lists.newArrayList();
+    private final List<String> dimensions;
+
+    private int fileIndex = 0;
+    private JsonParser jsonParser;
+    private Iterator<Row> it;
+
+    public ReplayableFirehose(InputRowParser parser) throws IOException
+    {
+      long counter = 0, totalBytes = 0, unparseable = 0, retryCount = 0;
+      Set<String> dimensionScratch = Sets.newHashSet();
+
+      File tmpDir = Files.createTempDir();
+      if (DELETE_TEMP_FILES) {
+        tmpDir.deleteOnExit();
```

Comment:
Deleting the directory in `ReplayableFirehose.close()` breaks the functionality of the firehose since it needs to retain the cache files between uses. We could delete the directory in `ReplayableFirehoseFactory.close()`, but I don't think it really makes a difference deleting the directory immediately vs when the process exits.